### ISSUE TITLE
SignalR lifecycle - reconnect/thread order

### ISSUE THREAD

**Issue by larslarsson** [SignalR/SignalR#4395]
How to manage client session between server & client using C# library in a correct thread safety way. E.g. after connection is correctly establish continue doing something else. If you freeze the thread on the first line of OnReconnect on the server, the clients states is transitioned from Reconnecting -> Connected & from my perspective not thread safe/ordering.

### Expected  behavior
Server reconnect code to be exectued before client lib changes status to connected.

### Actual behavior
Client is connected before OnReconnect on server has finished executing & can thus create threading issues/client session management issues.

**Comment by anurse** [comment]
> Server reconnect code to be exectued before client lib changes status to connected.

This is not guaranteed behavior. It's not feasible to synchronize this across the machine boundary between this client and server. The client and server are completely separate.

What threading issues are you seeing? Even though `OnReconnect` has fired, the server should process incoming data from the client successfully.

**Comment by larslarsson** [comment]
I am seeing other calls that client does after "connected" can execute before OnReconnect e.g. freeze it & the other calls are not queued for example.

**Comment by larslarsson** [comment]
Further, if you restart the server - the client will "OnReconnected" (no passing connected) thus how can the server properly keep track of clients in a thread safe fashion - to me its a big race condition. My reconnect is usually - if client does not exists, handle as connected.

**Comment by anurse** [comment]
Can you provide a runnable sample that reproduces this (or since it is intermittent, even just a sample to illustrate the problem)? It's not clear to me what "freeze" you're referring to. What exact behavior are you seeing on the client that is unexpected? Exceptions? Hangs? And exactly where are you seeing them?

**Comment by larslarsson** [comment]
Freeze I mean freeze the thread invisual studio or. sleep or something to cause a delay/force race condition.
Server:
        public override async Task OnReconnected()
        { <!-- breakpoint here or Thread.Sleep(...)

Simple console client:
        private async void OnStateChanged(StateChange state)
        {
            this.consolePrintingMessages = false;
            Console.WriteLine($"[{DateTimeOffset.Now}] StateChanged: {state.OldState}->{state.NewState}");
            await this.syncRoot.WaitAsync();
            try
            {
                switch(state.NewState)
                {
                    case ConnectionState.Connected:
                        await this.Subscribe();
                        break;
                    default:
                        this.isSubscribed = false;
                        break;
                }
            }
            finally
            {
                this.syncRoot.Release();
            }
        }


If you require a complete sample, then I have to get back to you to make a clean client/dummy standalone server.

**Comment by anurse** [comment]
Which thread is freezing? The Client is freezing when trying to send a message while the server is blocked? I think a complete sample would be best here. Let us know when you're able to get back to us with a sample (even if it's just illustrative because I realize race conditions are hard to reproduce). Making it very clear which operations are hanging in which place (client or server) is important to understanding the issue.

In general, we haven't heard much in the way of customer issues for this kind of thread-safety issue, so I'd be surprised to see a problem here. Let's see what we can find out though! It may just be a misunderstanding or documentation issue that we can help clear up for you.

**Comment by larslarsson** [comment]
Sorry I did not mean performance freezing, the function to pause a thread in visual studio is called freeze thread. But just put breakpoint there as indicated & the log is still printed on client.

**Comment by anurse** [comment]
> the log is still printed on client & thus continue with subscribe

Yes, because the client isn't directly coupled to the server. The message should be queued by the server until it can process it. If the server is completely locked up, then the message will still be in the client's TCP buffer for retransmission. What **specific data loss or exception issue** are you seeing? A log on the client doesn't guarantee that the message was immediately processed on the server, it's only possible to know that for sure when a response is received, but SignalR logs when the message is *sent* not necessarily when it's *processed by the server*. If you unfreeze the threads on the server, it should continue processing incoming messages.

**Comment by larslarsson** [comment]
My appoligies for the wait.

[SignalRSample.zip](https://github.com/SignalR/SignalR/files/3582725/SignalRSample.zip)

Steps:
1) Start server (using SignalRSampleServer.sln)
2) Start client (using SignalRSampleClient.sln)
3) Wait for subscription complete
4) Restart server

Server result:
Attempt to listen to url: http://localhost:8080/signalr/
SignalR: /signalr/signalr/hubs
WebApi: //signalr/api
Web: /signalr/
Starting connection. Press Ctrl-C to close.
Constructor
OnReconnected: client: 7c444116-e578-427b-8d61-361d99a2235d
Constructor
Subscribe: client: 7c444116-e578-427b-8d61-361d99a2235d
[DONE] Subscribe: client: 7c444116-e578-427b-8d61-361d99a2235d
[DONE] OnReconnected: client: 7c444116-e578-427b-8d61-361d99a2235d

Client result:
Starting connection. Press Ctrl-C to close.
[2019-09-06 09:25:57 +02:00] connect begin
[2019-09-06 09:25:57 +02:00] StateChanged: Disconnected->Connecting
[2019-09-06 09:25:58 +02:00] StateChanged: Connecting->Connected
[2019-09-06 09:25:58 +02:00] Messages received: Server-OnConnected
[2019-09-06 09:25:58 +02:00] connect successfully
[2019-09-06 09:25:58 +02:00] Subscribe begin
[2019-09-06 09:25:58 +02:00] Messages received: Server-Subscribe
[2019-09-06 09:26:03 +02:00] Subscribe successfully
[2019-09-06 09:26:27 +02:00] Error: An internal WebSocket error occurred. Please see the innerException, if present, for more details. , data: {}
[2019-09-06 09:26:27 +02:00] StateChanged: Connected->Reconnecting
[2019-09-06 09:26:27 +02:00] Reconnecting
[2019-09-06 09:26:29 +02:00] StateChanged: Reconnecting->Connected
[2019-09-06 09:26:29 +02:00] Subscribe begin
[2019-09-06 09:26:29 +02:00] Reconnected
[2019-09-06 09:26:29 +02:00] Messages received: Server-Subscribe
[2019-09-06 09:26:34 +02:00] Subscribe successfully
[2019-09-06 09:26:59 +02:00] Messages received: Server-OnReconnected

**Comment by larslarsson** [comment]
Another peculiarity (no sample) with startup where reconnect get queued, maybe related to above. Or lack of understand of what the lifecycle can be used for. Don't let this post overshadow the above one thou.

```
2019-09-16` 15:42:40,908 [35    ] INFO  (RealtimeHub) - OnReconnected: client: <client1>, connection: d83f0e16-ac70-457e-822a-eba5600a8eae
2019-09-16 15:42:40,908 [34    ] INFO  (RealtimeHub) - OnReconnected: client: <client1>, connection: d83f0e16-ac70-457e-822a-eba5600a8eae
2019-09-16 15:42:40,908 [33    ] INFO  (RealtimeHub) - OnReconnected: client: <client1>, connection: d83f0e16-ac70-457e-822a-eba5600a8eae
2019-09-16 15:42:40,911 [17    ] INFO  (RealtimeHub) - OnReconnected: client: <client1>, connection: d83f0e16-ac70-457e-822a-eba5600a8eae
```

**Comment by anurse** [comment]
I'm still not clear on what behavior is unexpected for you here. Going back to your original post: It is *by design* that the server's `OnReconnected` handler may run **after** the client transitions back to being connected. There is nothing that rigidly couples the two state transitions together. However, the messages should be processed in order. OnReconnected is not a blocking event, messages can be processed while your code is still running.

Similarly, ASP.NET SignalR is resilient even to server-side restarts, which is why you see an `OnReconnected` event even when you restart the server (and might expect `OnConnected`). The connection ID was originally issued by the server when the reconnect comes in, the server is able to detect (even though it was restarted) that this connection is **not** a new connection but rather a reconnection (this is because the client actually issues a different kind of connection request when reconnecting).

I ran your sample as suggested and don't see any unexpected errors. When the server is stopped, the client disconnects, and when reconnected it starts back up and the Subscribe method is called again.

What is the _misbehavior_ you are seeing? It's *not expected* that you can exactly correlate actions on the client to actions on the server on the same time scale.

**Comment by larslarsson** [comment]
Is there is anything discussing the threading & ordering I should read that clarifies this? My expectation was that lifecycle hooks are part of state transitions & I presumed they would have to be completed before other messages can be process. Was quite some time ago I went throu the documentation & seems to have changed a bit over the years so I might be missing something, otherwise here are my points.

1) Find it inconsistent saying its not part of the same stream - the timeline [graphs](https://docs.microsoft.com/en-us/aspnet/signalr/overview/guide-to-the-api/handling-connection-lifetime-events#connection-lifetime-terminology-and-scenarios) marking OnReconnect on the same timeline as the other two lifecycle events & I no further explaination about any seperate threading.
2) Pushes issues down to library consumer, is it wrong to argue the [sample code ](https://docs.microsoft.com/en-us/aspnet/signalr/overview/guide-to-the-api/mapping-users-to-connections#IUserIdProvider) about tracking user in memory is not thread safe or consistant if adding a sleep to force a delay in the OnReconnect would make clients being able to send messages while SendChatMessage would be inconsistent with intended use case? This is without discussing if the loop is even thread safe or why inner lock. 
3) What can OnReconnect be used for? What does people generally use it for? I can only see logging as a usage right now?

(2) is what I tried to do & creating a state around the client & threw exception when it does not exist 

Not part of the core issue but is multiple OnReconnect in short interval & parallel (different threads) to be expected? Not tested or  investigated but can ["You won't see the sequence OnConnected, OnDisconnected, OnReconnected for a given connection." ](https://docs.microsoft.com/en-us/aspnet/signalr/overview/guide-to-the-api/hubs-api-guide-server#asyncgroupmethods) be guaranteed if OnReconnected is run seperatly from the main threading?

Sorry for long & probably a confusing post.

**Comment by larslarsson** [comment]
Did I have a valid point or is my post/arguments impossible to understand?
Summarizing my post would be: documentation & samples does not align with current implementation & reconnet is useless if not part of same thread (safety) as connect & disconnect.

**Comment by BrennanConroy** [comment]
Hi. It looks like this is a question about how to use SignalR. This repo is in maintenance mode and we only have the resources to fix major issues, we are not a general-purpose forum. To get a faster response we suggest posting your questions to [StackOverflow](https://www.stackoverflow.com/).
